### PR TITLE
Implement `@EnsuresPresentIf` postcondition annotation

### PR DESCRIPTION
**Pull request by jyoo980**

### CODE REVIEW

**Comment by jyoo980** [comment]
Ready for review once I write some tests.

File: checker/tests/optional/EnsuresPresentIfTest.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  }
+
+  @EnsuresPresentIf(result = true, expression = "getOptId()")
+  public boolean hasPresentId1() {
+    return getOptId().isPresent();
+  }
+
+  void client() {
+    if (hasPresentId1()) {
+      optId.get();
```
Comment (jyoo980) [inline checker/tests/optional/EnsuresPresentIfTest.java:21]: @mernst this line reports `error: (method.invocation)` but I am not sure why. I think I am annotating `hasPresentId1` correctly.

Comment (mernst) [reply checker/tests/optional/EnsuresPresentIfTest.java:21]: This line should have been `getOptId().get()`.  Then the annotation works.  I added more test code to illustrate.

File: docs/manual/advanced-features.tex [hunk 2]
```diff
@@ -1233,6 +1233,8 @@
 \item \refqualclass{checker/nullness/qual}{RequiresNonNull}
 \item \refqualclass{checker/nullness/qual}{EnsuresNonNull}
 \item \refqualclass{checker/nullness/qual}{EnsuresNonNullIf}
+\item \refqualclass{checker/optional/qual}{EnsuresPresent}
```
Comment (mernst) [inline docs/manual/advanced-features.tex:1236]: I suggest adding `RequiresPresent`, for completeness.  That can be in a separate pull request.

File: checker/tests/optional/EnsuresPresentIfTest.java (hunk 1 above)
Comment (jyoo980) [reply checker/tests/optional/EnsuresPresentIfTest.java:80]: Right, that makes sense. This test would have passed if I instead wrote
```
expression="optId"
```
Right?

**Review by mernst — APPROVED** [review]
Thanks!

File: checker/tests/optional/EnsuresPresentIfTest.java (hunk 1 above)
Comment (mernst) [reply checker/tests/optional/EnsuresPresentIfTest.java:80]: Correct.  I added such a test.

File: checker-qual/src/main/java/org/checkerframework/checker/optional/qual/EnsuresPresentIf.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.checkerframework.framework.qual.ConditionalPostconditionAnnotation;
+import org.checkerframework.framework.qual.InheritedAnnotation;
+
+/**
+ * Indicates that the given expressions of type Optional&lt;T&gt; are present, if the method returns
+ * the given result (either true or false).FOO
```
Comment (jyoo980) [inline checker-qual/src/main/java/org/checkerframework/checker/optional/qual/EnsuresPresentIf.java:13]: @mernst Did you mean to write `.FOO` here?

Comment (jyoo980) [reply checker-qual/src/main/java/org/checkerframework/checker/optional/qual/EnsuresPresentIf.java:13]: Addressed in  #6269
